scripts/tools/memory/memory_report.py

```diff
@@ -0,0 +1,286 @@
+
+
+def generate_csv_report(results: dict[str, tuple[int, int, str]], csv_filename: str):
+    """Generates a CSV report of the memory usage results.
+
+    Args:
+        results: A dictionary mapping target names to (flash, ram, details) tuples.
+        csv_filename: The output CSV filename.
+    """
+    try:
+        with open(csv_filename, 'w', newline='') as f:  # Use newline='' for correct CSV handling
+            writer = csv.writer(f)
+            writer.writerow(["Application", "FLASH (bytes)", "RAM (bytes)", "Details"])
+            for app, (flash, ram, details) in results.items():
+                writer.writerow([app, flash, ram, details])
+        print(f"CSV Memory summary saved to {csv_filename}")
+    except Exception as e:
+        print(f"Error generating CSV report: {e}")
+
+
+def generate_html_report(csv_file_path: str, html_page_title: str, html_table_title: str, html_out_dir: str, sha: str):
+    """Generates an HTML report from a CSV file, inlined."""
+    try:
+        now = datetime.now().strftime("%d/%m/%Y %H:%M:%S")
+        html_report = f"""
+        <!DOCTYPE html>
+        <html>
+        <head>
+        <style>
+        h1 {{
+          font-family: Tahoma, Geneva, sans-serif;
+          font-size: 32px; color: #333;
+          text-align: center;
+        }}
+        h2 {{
+          font-family: Tahoma, Geneva, sans-serif;
+          font-size: 22px; color: #333;
+          text-align: center;
+        }}
+        h4 {{
+          font-family: Tahoma, Geneva, sans-serif;
+          font-size: 14px; color: #333;
+          text-align: left;
+        }}
+        table {{
+          border-collapse: collapse;
+          font-family: Tahoma, Geneva, sans-serif;
+          margin-left: auto;
+          margin-right: auto;
+          width: 80%;
+        }}
+        table td {{
+          padding: 15px;
+        }}
+        td[value="FAIL"] {{
+          color: red;
+        }}
+        td[value="PASS"] {{
+          color: green;
+        }}
+        th {{
+          background-color: #54585d;
+          color: #ffffff;
+          font-weight: bold;
+          font-size: 15px;
+          border: 1px solid #54585d;
+        }}
+        table tbody td {{
+          color: #636363;
+          border: 1px solid #dddfe1;
+        }}
+        table tbody tr {{
+          background-color: #f9fafb;
+        }}
+        table tbody tr:nth-child(odd) {{
+          background-color: #ffffff;
+        }}
+        </style>
+        <title>{html_page_title}</title>
+        </head>
+        <body>
+        <h1>{html_page_title}</h1>
+        <hr>
+        <h4>Generated on: {now}<br>SHA: {sha}</h4>
+        <hr>
+        """
+
+        table = pandas.read_csv(csv_file_path)
+        table['Details'] = "<details><summary>Show/Hide</summary>" + \
+            table['Details'].str.replace('\n', '<br>') + "</details>"
+
+        html_table = f"<h2>{html_table_title}</h2>"
+        html_table += table.to_html(escape=False)
```

Comment:
I still believe we should separate out the templates into their own files instead of interleaving strings / constructing strings.